velox/expression/EvalCtx.h

```diff
@@ -361,12 +361,11 @@ class EvalCtx {
   ErrorVectorPtr errors_;
 };
 
-/// Utility wrapper struct that is used to temporarily reset the value of the an
-/// ExprCtx till it goes out of scope. EvalCtx::saveAndReset() is used to
-/// achieve that. The old context can also be explicitly restored using
-/// EvalCtx::restore().
-struct ScopedContextSaver {
-  ~ScopedContextSaver();
+/// Utility wrapper struct that is used to temporarily reset the value of the
+/// EvalCtx. EvalCtx::saveAndReset() is used to achieve that. The old context
+/// must be explicitly restored using restoreContext or EvalCtx::restore().
+struct ContextSaver {
+  void restoreContext();
```

Comment:
This is error prone as you already pointed out.  Since we cannot use it as RAII and does not plan to restore in case of exception here, how about wrapping the scope in a functor?  Something like:

```c++
/// Only restore the context when the body executes succesfully.
template <typename F>
auto withContextSaver(F f) {
  ContextSaver saver;
  auto res = f(saver);
  if (saver.context) {
    saver.context->restore(saver);
  }
  return res;
}
```

Reply:
Thanks! That's a good idea, the only tweak I made was treating F as void for now since we don't need to return anything and I didn't want to put "return nullptr;" everywhere.

We can always add a withContextSaverAndReturn or something when it's needed.